src/compiler/parser.ts

```diff
@@ -521,16 +521,17 @@ namespace ts {
         let parseErrorBeforeNextFinishedNode: boolean = false;
 
         export function parseSourceFile(fileName: string, _sourceText: string, languageVersion: ScriptTarget, _syntaxCursor: IncrementalParser.SyntaxCursor, setParentNodes?: boolean): SourceFile {
-            initializeState(fileName, _sourceText, languageVersion, _syntaxCursor);
+            const isJavaScriptFile = hasJavaScriptFileExtension(fileName) || _sourceText.lastIndexOf("// @language=javascript", 0) === 0;

```

Comment:
Wouldn't it be more efficient to use `substr` to just get the first N characters of the text and then compare to the magic string? With `lastIndexOf` presumably the entire string is being scanned.